### PR TITLE
[SPARK-13511][SQL] Add wholestage codegen for limit

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-13511
## What changes were proposed in this pull request?

Current limit operator doesn't support wholestage codegen. This is open to add support for it.

In the `doConsume` of `GlobalLimit` and `LocalLimit`, we use a count term to count the processed rows. Once the row numbers catches the limit number, we set the variable `stopEarly` of `BufferedRowIterator` newly added in this pr to `true` that indicates we want to stop processing remaining rows. Then when the wholestage codegen framework checks `shouldStop()`, it will stop the processing of the row iterator.

Before this, the executed plan for a query `sqlContext.range(N).limit(100).groupBy().sum()` is:

```
TungstenAggregate(key=[], functions=[(sum(id#5L),mode=Final,isDistinct=false)], output=[sum(id)#6L])
+- TungstenAggregate(key=[], functions=[(sum(id#5L),mode=Partial,isDistinct=false)], output=[sum#9L])
   +- GlobalLimit 100
      +- Exchange SinglePartition, None
         +- LocalLimit 100
            +- Range 0, 1, 1, 524288000, [id#5L]
```

After add wholestage codegen support:

```
WholeStageCodegen
:  +- TungstenAggregate(key=[], functions=[(sum(id#40L),mode=Final,isDistinct=false)], output=[sum(id)#41L])
:     +- TungstenAggregate(key=[], functions=[(sum(id#40L),mode=Partial,isDistinct=false)], output=[sum#44L])
:        +- GlobalLimit 100
:           +- INPUT
+- Exchange SinglePartition, None
   +- WholeStageCodegen
      :  +- LocalLimit 100
      :     +- Range 0, 1, 1, 524288000, [id#40L]
```
## How was this patch tested?

A test is added into BenchmarkWholeStageCodegen.

### CODE REVIEW

**Comment by viirya** [comment]
Just realized this implementation is wrong. Will fix it later.

**Comment by hvanhovell** [comment]
Mark it as WIP?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 1]
```diff
@@ -43,12 +44,33 @@ case class CollectLimit(limit: Int, child: SparkPlan) extends UnaryNode {
...
+  override def doConsume(ctx: CodegenContext, input: Seq[ExprCode]): String = {
+    val countTerm = ctx.freshName("count")
+    ctx.addMutableState("int", countTerm, s"$countTerm = 0;")
+    ctx.currentVars = input
+    s"""
+       | while ($countTerm < $limit) {
+       |   $countTerm += 1;
+       |   ${consume(ctx, ctx.currentVars)}
+       | }
+       | if (true) return;
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: This is special?

**Comment by rxin** [comment]
Can you update the pr description to describe how you are supporting it?

**Comment by viirya** [comment]
The failed test `ParquetHadoopFsRelationSuite` is due to the lack of short type support in `UnsafeRowParquetRecordReader`. I submitted another PR #11412 to fix it. The change is also included here to prove the tests can be passed.

**Comment by viirya** [comment]
cc @davies

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java [hunk 2]
```diff
@@ -765,6 +765,9 @@ private void readIntBatch(int rowId, int num, ColumnVector column) throws IOExce
       } else if (DecimalType.is64BitDecimalType(column.dataType())) {
         defColumn.readIntsAsLongs(
             num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
+      } else if (column.dataType() == DataTypes.ShortType) {
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java]: Could you pull this out as an separate PR? so we can merge it quickly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 3]
```diff
@@ -56,6 +57,28 @@ trait BaseLimit extends UnaryNode {
...
+  }
+
+  protected override def doProduce(ctx: CodegenContext): String = {
+    child.asInstanceOf[CodegenSupport].produce(ctx, this)
+  }
+
+  override def doConsume(ctx: CodegenContext, input: Seq[ExprCode]): String = {
+    val countTerm = ctx.freshName("count")
+    ctx.addMutableState("int", countTerm, s"$countTerm = 0;")
+    ctx.currentVars = input
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: We don't need to update `currentVars` here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 4]
```diff
@@ -70,6 +70,20 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
     */
   }
 
+  ignore("range/limit/sum") {
+    val N = 500 << 20
+    runBenchmark("range/limit/sum", N) {
+      sqlContext.range(N).limit(100).groupBy().sum().collect()
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: For 100 rows, we may can't really see the performance difference, could you increase it to 1 million?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java [hunk 5]
```diff
@@ -64,7 +66,16 @@ protected void append(InternalRow row) {
    * If it returns true, the caller should exit the loop (return from processNext()).
    */
   protected boolean shouldStop() {
-    return !currentRows.isEmpty();
+    return !currentRows.isEmpty() || stopEarly;
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java]: I think we could only override `shouldStop()` for Limit, so we won't add a branch when we don't have Limit.

```
ctx.addFunction("shouldStop", """
  protected boolean shouldStop() {
    return !currentRows.isEmpty() || stopEarly;
 }""")
```

Please be care full that we can't have two `shouldStop()` in generated class.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java]: Ah, it has been merged, let me sync with it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java [hunk 6]
```diff
@@ -35,6 +35,8 @@
   // used when there is no column in output
   protected UnsafeRow unsafeRow = new UnsafeRow(0);
 
+  protected boolean stopEarly = false;
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java]: Since `stopEarly` is only accessed generated functions, we don't need this anymore.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java]: We could use `addMutableState`

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java]: yeah. I am updating it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 7]
```diff
@@ -78,4 +78,21 @@ class WholeStageCodegenSuite extends SparkPlanTest with SharedSQLContext {
...
     assert(df.collect() === Array(Row(1), Row(2), Row(3)))
   }
+
+  test("Limit should be included in WholeStageCodegen") {
+    val df = sqlContext.range(10000).limit(100).sort(col("id"))
+    val plan = df.queryExecution.executedPlan
+
+    assert(plan.find(p =>
+      p.isInstanceOf[WholeStageCodegen] &&
+        p.asInstanceOf[WholeStageCodegen].plan.isInstanceOf[Sort] &&
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: The sort is not related to limit, could you remove it from this PR? (we may revert the commit for sort)

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: Yeah, because we can't leave limit as last operator otherwise it will transform to collect limit, so I add a sort here. I will remove it once I am back to laptop (few hours later).

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: These kind of tests are easy to break, we may don't need this.

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: Agreed. Let me remove this later.

**Comment by davies** [comment]
LGTM, merging this into master, thanks!
